### ISSUE TITLE
SC2034 fails to understand scopes

### ISSUE THREAD

**Issue by stdedos** [koalaman/shellcheck#1497]
#### For bugs
- Rule Id (if any, e.g. SC1000): `SC2034`
- My shellcheck version (`shellcheck --version` or "online"): online + 0.6.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

Shellcheck somehow misses the scope of a variable, and figuring out all outstanding SC2034 can be a uninteresting, repetitive task. Follow along:

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

_comp1() {
	local cur opts
	local cur prev opts
	COMPREPLY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"
	opts="a b c"

	mapfile -t COMPREPLY <<< "$(compgen -W "${opts}" -- "${cur}")"
	return 0
}
complete -F _comp1 comp1

_comp2() {
	local cur opts
	local cur prev opts
	COMPREPLY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"
	opts="a b c"

	mapfile -t COMPREPLY <<< "$(compgen -W "${opts}" -- "${cur}")"
	return 0
}

complete -F _comp2 comp2
```

```bash
$ shellcheck myscript
 
Line 21:
        prev="${COMP_WORDS[COMP_CWORD-1]}"
        ^-- SC2034: prev appears unused. Verify use (or export if used externally).

$ 
```

```sh
#!/bin/bash

_comp1() {
	local cur opts
	local cur prev opts
	COMPREPLY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"
	opts="a b c"

	mapfile -t COMPREPLY <<< "$(compgen -W "${opts}" -- "${cur}")"
	return 0
}
complete -F _comp1 comp1

_comp2() {
	local cur opts
	local cur prev opts
	COMPREPLY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	# prev="${COMP_WORDS[COMP_CWORD-1]}"
	opts="a b c"

	mapfile -t COMPREPLY <<< "$(compgen -W "${opts}" -- "${cur}")"
	return 0
}

complete -F _comp2 comp2
```

```bash
$ shellcheck myscript
 
Line 18:
        local cur prev opts
                  ^-- SC2034: prev appears unused. Verify use (or export if used externally).

$ 
```

```sh
#!/bin/bash

_comp1() {
	local cur opts
	local cur prev opts
	COMPREPLY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	prev="${COMP_WORDS[COMP_CWORD-1]}"
	opts="a b c"

	mapfile -t COMPREPLY <<< "$(compgen -W "${opts}" -- "${cur}")"
	return 0
}
complete -F _comp1 comp1

_comp2() {
	local cur opts
	# local cur prev opts
	COMPREPLY=()
	cur="${COMP_WORDS[COMP_CWORD]}"
	# prev="${COMP_WORDS[COMP_CWORD-1]}"
	opts="a b c"

	mapfile -t COMPREPLY <<< "$(compgen -W "${opts}" -- "${cur}")"
	return 0
}

complete -F _comp2 comp2
```

```bash
$ shellcheck myscript
 
Line 8:
        prev="${COMP_WORDS[COMP_CWORD-1]}"
        ^-- SC2034: prev appears unused. Verify use (or export if used externally).

$ 
```

#### Here's what I wanted or expected to see:

```bash
$ shellcheck myscript
 
Line 8:
        prev="${COMP_WORDS[COMP_CWORD-1]}"
        ^-- SC2034: prev appears unused. Verify use (or export if used externally).
 
Line 21:
        prev="${COMP_WORDS[COMP_CWORD-1]}"
        ^-- SC2034: prev appears unused. Verify use (or export if used externally).

$ 
```

**Comment by rollf** [comment]
```bash
$ cat test.sh
#!/usr/bin/env bash

f() {
  local X # should warn about unused X but doesn't ....
  local Y
}

g() {
  local X # ... because the same name is used here. 
  X=3
  echo "$X"
}
$ shellcheck test.sh

In test.sh line 5:
  local Y
        ^-- SC2034 (warning): Y appears unused. Verify use (or export if used externally).

For more information:
  https://www.shellcheck.net/wiki/SC2034 -- Y appears unused. Verify use (or ...
```

**Comment by wileyhy** [comment]
I've noticed that, too.

One workaround is to scan section by section:
- comment out all
- uncomment section 1, scan, do fixes
- uncomment section 2, scan, do fixes
- rinse/repeat

Someone else probably knows a better way to do it.
